[Merged by Bors] - feat(linear_algebra/affine_space): affine independence

jsm28 opened the pull request:
Define affine independent indexed families of points (in terms of no
nontrivial `weighted_vsub` in the family being zero), prove that a
family of at most one point is affine independent, and prove a family
of points is affine independent if and only if a corresponding family
of subtractions is linearly independent.

There are of course other equivalent descriptions of affine
independent families it will be useful to add later (that the affine
span of all proper subfamilies is smaller than the affine span of the
whole family, that each point is not in the affine span of the rest;
in the case of a family indexed by a `fintype`, that the dimension of
the affine span is one less than the cardinality).  But I think the
definition and one equivalent formulation is a reasonable starting
point.


---
<!-- put comments you want to keep out of the PR commit here -->

jsm28 committed:
feat(linear_algebra/affine_space): affine independence

Define affine independent indexed families of points (in terms of no
nontrivial `weighted_vsub` in the family being zero), prove that a
family of at most one point is affine independent, and prove a family
of points is affine independent if and only if a corresponding family
of subtractions is linearly independent.

There are of course other equivalent descriptions of affine
independent families it will be useful to add later (that the affine
span of all proper subfamilies is smaller than the affine span of the
whole family, that each point is not in the affine span of the rest;
in the case of a family indexed by a `fintype`, that the dimension of
the affine span is one less than the cardinality).  But I think the
definition and one equivalent formulation is a reasonable starting
point.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine-independent

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine-independent

jsm28 committed:
Another big_operators lemma relevant to affine independence.

jsm28 committed:
Split out a couple of lemmas (more hopefully to follow).

jsm28 committed:
More big_operators lemmas.

jsm28 committed:
Move new lemmas to a more appropriate place.

kim-em commented on src/algebra/big_operators.lean:
```diff
@@ -338,6 +350,30 @@ by haveI := classical.dec_eq α; exact
 
+/-- A product over `s.subtype p` equals one over `s.filter p`. -/
+@[to_additive "A sum over `s.subtype p` equals one over `s.filter p`."]
+lemma prod_subtype_eq_prod_filter (f : α → β) {p : α → Prop} [decidable_pred p] :
+  ∏ x in s.subtype p, f x = ∏ x in s.filter p, f x :=
```
Perhaps this deserves `@[simp]`? It it removing a coercion.

jsm28 replied:
Added more use of `@[simp]` as suggested.

kim-em commented on src/data/finset.lean:
```diff
@@ -1357,6 +1357,37 @@ protected def subtype {α} (p : α → Prop) [decidable_pred p] (s : finset α)
+/-- `s.subtype p` converts back to `s.filter p` with
+`embedding.subtype`. -/
+lemma subtype_map (p : α → Prop) [decidable_pred p] :
+  (s.subtype p).map (function.embedding.subtype _) = s.filter p :=
```
`@[simp]`?

jsm28 replied:
Added more use of `@[simp]` as suggested.

kim-em commented on src/data/finset.lean:
```diff
@@ -1357,6 +1357,37 @@ protected def subtype {α} (p : α → Prop) [decidable_pred p] (s : finset α)
+
+/-- If all elements of a `finset` satisfy the predicate `p`,
+`s.subtype p` converts back to `s` with `embedding.subtype`. -/
+lemma subtype_map_of_mem {p : α → Prop} [decidable_pred p] (h : ∀ x ∈ s, p x) :
+  (s.subtype p).map (function.embedding.subtype _) = s :=
```
`@[simp]`?

jsm28 replied:
Added more use of `@[simp]` as suggested.

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -164,6 +165,26 @@ begin
+/-- The weighted sum is unaffected by removing the base point, if
+present, from the set of points. -/
+lemma weighted_vsub_of_point_erase (w : ι → k) (p : ι → P) (i : ι) :
+  (s.erase i).weighted_vsub_of_point V p (p i) w = s.weighted_vsub_of_point V p (p i) w :=
```
```suggestion
@[simp] lemma weighted_vsub_of_point_erase (w : ι → k) (p : ι → P) (i : ι) :
  (s.erase i).weighted_vsub_of_point V p (p i) w = s.weighted_vsub_of_point V p (p i) w :=
```

kim-em commented on src/linear_algebra/affine_space.lean:
```diff
@@ -164,6 +165,26 @@ begin
+/-- The weighted sum is unaffected by adding the base point, whether
+or not present, to the set of points. -/
+lemma weighted_vsub_of_point_insert (w : ι → k) (p : ι → P) (i : ι) :
+  (insert i s).weighted_vsub_of_point V p (p i) w = s.weighted_vsub_of_point V p (p i) w :=
```
```suggestion
@[simp] lemma weighted_vsub_of_point_insert (w : ι → k) (p : ι → P) (i : ι) :
  (insert i s).weighted_vsub_of_point V p (p i) w = s.weighted_vsub_of_point V p (p i) w :=
```

kim-em commented on src/algebra/big_operators.lean:
```diff
@@ -338,6 +350,30 @@ by haveI := classical.dec_eq α; exact
+over `s.subtype p` equals that product over `s`. -/
+@[to_additive "If all elements of a `finset` satisfy the predicate `p`, a sum
+over `s.subtype p` equals that sum over `s`."]
+lemma prod_subtype_of_mem (f : α → β) {p : α → Prop} [decidable_pred p]
+    (h : ∀ x ∈ s, p x) : ∏ x in s.subtype p, f x = ∏ x in s, f x :=
```
You've added this lemma, but then not used it. Are you still planning on refactoring the big proof further?

jsm28 replied:
Changing anything in `algebra.big_operators` or `data.finset` means Lean wants to recompile large parts of mathlib. So my workflow for adding lemmas there was to write those lemmas and their users elsewhere, then move them to the right place, then push those lemmas and go away and wait for CI to recompile mathlib before seeing if the code to use the lemmas still works with the final version of the lemmas in the correct place (and only then pushing the code using the lemmas if it does work). I've now pushed the refactor to use those lemmas (but still need to refactor the first half of the proof, which could involve adding more such lemmas).

jsm28 committed:
Use new lemmas in proof.

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

jsm28 committed:
Update src/linear_algebra/affine_space.lean

Co-authored-by: Scott Morrison <scott.morrison@gmail.com>

jsm28 committed:
Add another big_operators lemma.

jsm28 committed:
Add more use of @[simp].

jsm28 committed:
Use finset.sum_subtype_map_embedding.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine-independent

jsm28 committed:
Remove @[simp] annotation that fails linting.

jsm28 committed:
Remove another @[simp] annotation that fails linting.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine-independent

jsm28 committed:
Remove another @[simp] that fails linting.

jsm28 committed:
Add another finset.filter lemma, suggested by simp linter failures.

jsm28 committed:
simp linter reports simp can now prove filter_true

So remove simp attribute from it.  Reported proof:

by simp only [finset.filter_true_of_mem, finset.filter_congr_decidable, forall_true_iff]

jcommelin commented on src/data/finset.lean:
```diff
@@ -844,7 +844,7 @@ theorem filter_filter (s : finset α) :
 ext $ assume a, by simp only [mem_filter, and_comm, and.left_comm]
 
-@[simp] lemma filter_true {s : finset α} [h : decidable_pred (λ _, true)] :
+lemma filter_true {s : finset α} [h : decidable_pred (λ _, true)] :
```
Can you explain this change?

jsm28 replied:
The linter reports that `simp` can prove this lemma using the new `filter_true_of_mem`: `by simp only [finset.filter_true_of_mem, finset.filter_congr_decidable, forall_true_iff]`. Is there some better fix for the linter warning in that case than removing the `@[simp]` from the lemma that `simp` can prove?

fpvandoorn replied:
That is the correct fix.

jcommelin replied:
@jsm28 Thanks! I didn't expect that `simp` could already prove this.

jsm28 committed:
Factor out one more finset lemma.

fpvandoorn commented on src/linear_algebra/affine_space.lean:
```diff
@@ -212,6 +233,77 @@ end
+  { intro h,
+    rw linear_independent_iff',
+    intros s g hg i hi,
+    set f : ι → k := λ x, if hx : x = i1 then -∑ y in s, g y else g ⟨x, hx⟩ with hfdef,
```
I think you can define `f` in terms of `function.update`. That might simplify some properties of `f`.

fpvandoorn replied:
Oh wait, `g` depends on the proof `hx`. Disregard this.

ChrisHughes24 commented on src/linear_algebra/affine_space.lean:
```diff
@@ -212,6 +233,77 @@ end
+
+/-- A family is affinely independent if and only if the differences
+from a base point in that family are linearly independent. -/
+lemma affine_independent_iff_linear_independent_vsub (p : ι → P) (i1 : ι) :
```
Would this proof be easier if things like `weighted_vsub` used `finsupp` instead of a pair of `ι → k` and `s : finset ι`? Potentially there should be a definition of a free affine space, and affine independence should be defined in terms of this. Do we have morphisms of affine spaces? I don't have clear understanding of this though. @jcommelin any thoughts?

jsm28 replied:
I think it's natural for `weighted_vsub` to be defined in terms of sums over a `finset` since that's how all the lemmas in `big_operators` are defined. And just as it was concluded in #3122 not to have duplicate copies of the API for the case of summing over `finset.univ`, so it seems best not to duplicate it for sums over `finsupp.support` either. (I do expect to end up with at least one lemma involving `finsupp`: for all points in the affine span of an affine independent family, there exists a unique `finsupp` with sum 1 whose `affine_combination` is that point, i.e. a unique choice of barycentric coordinates for that point.)

jcommelin replied:
Yup, I don't like `finsupp.sum` etc... I would be in favour of removing that, because the API for `finset.sum` is a lot better.

jsm28 committed:
Merge remote-tracking branch 'origin/master' into affine-independent

PatrickMassot commented on src/linear_algebra/affine_space.lean:
```diff
@@ -212,6 +233,77 @@ end
+nontrivial weighted subtractions (where the sum of weights is 0) are
+0. -/
+def affine_independent (p : ι → P) : Prop :=
+∀ (s : finset ι) (w : ι → k) (h : ∑ i in s, w i = 0), s.weighted_vsub V p w = 0 → ∀ i ∈ s, w i = 0
```
```suggestion
∀ (s : finset ι) (w : ι → k), ∑ i in s, w i = 0 → s.weighted_vsub V p w = 0 → ∀ i ∈ s, w i = 0
```

PatrickMassot committed:
Update src/linear_algebra/affine_space.lean
